### PR TITLE
KAFKA-12446: Call subtractor before adder if key is the same

### PR DESCRIPTION
**Pull request by fqaiser94**
For full details, please see [KIP-904](https://cwiki.apache.org/confluence/x/P5VbDg). 

# Problem
During `KTable.groupBy`, we write into a repartition topic. Since the grouping key can change, we need to send separate events for the oldValue and the newValue to downstream nodes (where they will be “subtracted” and “added” respectively from/to the aggregate for the old key and the new key respectively). 

However, sending the oldValue and the newValue as separate events is not strictly necessary when the grouping key does not change and doing so poses two challenges for users: 
1. Firstly, the resulting KTable (i.e. the result of `KTable.groupBy(???).aggregate(???)`) can briefly be in an “inconsistent” state where the oldValue has been “subtracted” from the aggregate for the key but the newValue has not yet been “added” to the aggregate of the key because each event (oldValue, newValue) is processed separately.  
2. Secondly, if users fail to correctly configure their producers correctly to avoid re-ordering during `send()`, it’s possible the newValue may be sent (and therefore processed by the aggregator) before the oldValue. If the user’s`adder` and `subtractor` functions are non-commutative, this would put the aggregate in a permanently “inconsistent” state. 

Whilst there are ways to get around this issue by dropping down to the Processor API level, it would be nicer if this was handled by Kafka Streams more seamlessly. 

# Solution
If the grouping key has not changed, the oldValue and newValue events are guaranteed to be processed by the same processor.  As such, we should be able to send them as a single `Change<T>` event. The subtractor and adder functions can then be executed (in that order) and the KTable can be updated in a single “atomic” operation. In this way, we are able to remove any possibility of ending up in an “inconsistent” state. Also, note that sending the oldValue and newValue in the same event ensures that they can’t be re-ordered relative to each other irrespective of how a user has configured the producer settings. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 1]
```diff
@@ -45,34 +45,30 @@ public void setIfUnset(final Serializer<Void> defaultKeySerializer, final Serial
...
+
+            final int newDataLength = newData.length;
+            final int capacity = NEW_DATA_LENGTH_BYTES_SIZE + newDataLength + oldData.length;
+
+            return ByteBuffer
+                    .allocate(capacity)
+                    .putInt(newDataLength)
+                    .put(newData)
+                    .put(oldData)
+                    .array();
```
Comment (fqaiser94) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: With the change to the `ChangedSerializer` and `ChangedDeserializer` classes, I don’t think users will be able to just upgrade from a previous version of Kafka Streams easily. Any "inflight" messages written by older library versions will fail to deserialize correctly after the upgrade. 

Not sure how these types of “breaking” changes are typically handled. 

1. Is it simply a matter of noting this in the relevant upgrade doc i.e. users need to do an application-reset? 
2. Or do we want to write more code to handle upgrade scenarios? 
3. Or find a more backwards compatible way of writing this serde?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java [hunk 2]
```diff
@@ -90,14 +90,19 @@ public void process(final K key, final Change<V> change) {
 
             // if the selected repartition key or value is null, skip
             // forward oldPair first, to be consistent with reduce and aggregate
-            if (oldPair != null && oldPair.key != null && oldPair.value != null) {
-                context().forward(oldPair.key, new Change<>(null, oldPair.value));
+            final boolean oldPairNotNull = oldPair != null && oldPair.key != null && oldPair.value != null;
+            final boolean newPairNotNull = newPair != null && newPair.key != null && newPair.value != null;
+            if (oldPairNotNull && newPairNotNull && oldPair.key == newPair.key) {
```
Comment (fqaiser94) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java]: As noted by Matthias on the mailing list thread, this fix depends on a correct implementation of `.equals()` method for the key type. Would we need to document this assumption somewhere for users?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 3]
```diff
@@ -45,34 +45,30 @@ public void setIfUnset(final Serializer<Void> defaultKeySerializer, final Serial
...
+        // both old and new values cannot be null
+        if (oldValueIsNull && newValueIsNull) {
+            throw new StreamsException("Both old and new values are null in ChangeSerializer, which is not allowed.");
         } else {
-            if (data.oldValue == null) {
-                throw new StreamsException("Both old and new values are null in ChangeSerializer, which is not allowed.");
-            }
-
-            serializedKey = inner.serialize(topic, headers, data.oldValue);
+            final byte[] newData = newValueIsNull ? new byte[0] : inner.serialize(topic, headers, data.newValue);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]:
```suggestion
            final byte[] newData = newValueIsNull ? null : inner.serialize(topic, headers, data.newValue);
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 4]
```diff
@@ -45,34 +45,30 @@ public void setIfUnset(final Serializer<Void> defaultKeySerializer, final Serial
...
+            final byte[] oldData = oldValueIsNull ? new byte[0] : inner.serialize(topic, headers, data.oldValue);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]:
```suggestion
            final byte[] oldData = oldValueIsNull ? null : inner.serialize(topic, headers, data.oldValue);
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java [hunk 5]
```diff
@@ -45,16 +46,17 @@ public void setIfUnset(final Deserializer<Void> defaultKeyDeserializer, final De
...
-
-        System.arraycopy(data, 0, bytes, 0, bytes.length);
-
-        if (ByteBuffer.wrap(data).get(data.length - NEWFLAG_SIZE) != 0) {
-            return new Change<>(inner.deserialize(topic, headers, bytes), null);
-        } else {
-            return new Change<>(null, inner.deserialize(topic, headers, bytes));
-        }
+        return new Change<>(
+                newDataLength > 0 ? inner.deserialize(topic, headers, newData) : null,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java]: And empty BytesBuffer and a `null` is not the same, and we should distinguish between both. Cf my other comment in the serializer.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java]: `oldPair.key == newPair.key` is not using `equals()` but it does an object reference comparison. We need to use `oldPair.key.equals(newPiar.key)` (and test that `oldPair.key != null` to avoid a NPE.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java [hunk 6]
```diff
@@ -18,13 +18,14 @@
...
 import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serdes;
 
 import java.nio.ByteBuffer;
 import java.util.Objects;
 
 public class ChangedDeserializer<T> implements Deserializer<Change<T>>, WrappingNullableDeserializer<Change<T>, Void, T> {
 
-    private static final int NEWFLAG_SIZE = 1;
+    private static final int NEW_DATA_LENGTH_BYTES_SIZE = Integer.BYTES;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java]: Why do we rename this? (I am just wondering what this is actually...)

File: docs/streams/developer-guide/dsl-api.html [hunk 7]
```diff
@@ -1050,8 +1050,8 @@ <h4 class="anchor-heading"><a id="streams_concepts_globalktable" class="anchor-l
                                         more than once for a key as a result of having received input tombstone records for that key (see below).</li>
                                     <li>When the first non-<code class="docutils literal"><span class="pre">null</span></code> value is received for a key (e.g.,  INSERT), then only the adder is called.</li>
                                     <li>When subsequent non-<code class="docutils literal"><span class="pre">null</span></code> values are received for a key (e.g.,  UPDATE), then (1) the subtractor is
-                                        called with the old value as stored in the table and (2) the adder is called with the new value of the
-                                        input record that was just received.  The order of execution for the subtractor and adder is not defined.</li>
+                                        first called with the old value as stored in the table and then (2) the adder is called with the new value of the
+                                        input record that was just received.</li>
```
Comment (mjsax) [inline docs/streams/developer-guide/dsl-api.html]: Well, for the regular case in which the key changes, the order is still not defined...

We would need to be more precise and say:
> If and only if the extracted grouping key of the old and new value is the same, the subtractor will be called before the adder. The detection of this case depends on the correct implementation of the <code>equals()</code> method of the extracted key type.

**Review by mjsax — COMMENTED** [review]
I am just working through the review backlog and re-discovered this PR. Left a few comments. If you are still interested in working on this, happy to help you getting it merged.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java [hunk 8]
```diff
@@ -24,8 +24,6 @@
 
 public class ChangedDeserializer<T> implements Deserializer<Change<T>>, WrappingNullableDeserializer<Change<T>, Void, T> {
 
-    private static final int NEWFLAG_SIZE = 1;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java:27]: Ah. Re-reading the code I remember... We only encode the new or the old value in the bytes, and add an additional flag at the end, to indicate if it is the new or old value... Guess we should keep this variable?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java [hunk 9]
```diff
@@ -44,18 +42,30 @@ public void setIfUnset(final SerdeGetter getter) {
...
+            final int dataLength = buffer.getInt();
+            data = new byte[dataLength];
+            buffer.get(data);
         }
+        return data;
+    }
+
+    @Override
+    public Change<T> deserialize(final String topic, final Headers headers, final byte[] data) {
+        // The format we need to deserialize is:
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java:51]: For backward compatibility, we cannot just change the format. We need to keep the old format as `{BYTE_ARRAY oldOrNewValue}{BYTE oldNewFlag}` with `oldNewFlag == 0` if the encoded bytes represent the old value (and the the new value is `null`) and `oldNewFlag == 1` if the encoded bytes represent the new value (and thus the old value is `null`).

I guess, we can extend the format to add `oldNewFlag = 2` implying that we have serialized both old and new value in the byte array using format: `{INT newDataLenght}{BYTE_ARRAY newValue}{BYTE_ARRAY oldValue}{INT newOldFlag=2}` ?

(Note, we don't need to store old value length, as we can compute it as `data.length - newDataLength -1`.

Ie, we need to handle tree cases now... The two old/existing ones, plus the new one. This actually also raises one more question about backward compatibility -- if we update a single KS instance, and it starts writing the new format, there might still be an old instance that does not understand the new format... Thus, I think we need do a two round rolling bounce upgrade... In the first round, we still use only the old formats, and only update the version. In the second round, we change a config to enable the new format. I think we can re-use the existing `upgrade_from` config. However, for this kind of change, we would actually need a KIP... Are you familiar with the KIP process?

**Review by mjsax — COMMENTED** [review]
Thanks for updating the PR. -- Thinking about it once more, I actually think we need a KIP for this change, because of backward compatibility considerations. Should not be a complex KIP, but I don't feel confident merging this change without an approved KIP.

**Comment by mjsax** [comment]
@fqaiser94 -- are you interested in doing a KIP do move this forward?

**Comment by fqaiser94** [comment]
Heya @mjsax, apologies I've been travelling recently so I haven't been able to respond to any of your messages 😓 
Yes, I would be interested in doing a KIP to move this forward but I was hoping to address your remaining comments first (think I can do that this weekend). 
Does that make sense?

**Comment by mjsax** [comment]
Great! Sure, this works.

Looking forward to your KIP -- let me know if you have any question about the KIP process.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java (hunk 9 above)
Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java:51]: > For backward compatibility, we cannot just change the format.

Agreed, this was actually my first [concern](https://github.com/apache/kafka/pull/10747#discussion_r637657442) when I raised the initial PR XD

> I guess, we can extend the format to add oldNewFlag = 2 

Sounds good, implemented that in my latest changes. 

> Thus, I think we need do a two round rolling bounce upgrade... In the first round, we still use only the old formats, and only update the version. In the second round, we change a config to enable the new format. I think we can re-use the existing upgrade_from config. 

Will need to think about this a little more but I _think_ this makes sense to me. 
Would the code changes required to enable this look similar to what's already been done in `SubscriptionWrapperSerde`? e.g.  https://github.com/apache/kafka/blob/8d2e157b37902055bd5a8f4bd0f6ac29080910eb/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java#L68

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 10]
```diff
@@ -46,33 +45,40 @@ public void setIfUnset(final SerdeGetter getter) {
...
-
-            serializedKey = inner.serialize(topic, headers, data.newValue);
+        final boolean oldValueIsNull = data.oldValue == null;
+        final boolean newValueIsNull = data.newValue == null;
+
+        final byte[] newData = inner.serialize(topic, headers, data.newValue);
+        final byte[] oldData = inner.serialize(topic, headers, data.oldValue);
+
+        final int newDataLength = newValueIsNull ? 0 : newData.length;
+        final int oldDataLength = oldValueIsNull ? 0 : oldData.length;
```
Comment (fqaiser94) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: @mjsax One thing to note with this logic. if `inner.serialize` returns `null` for a non-null `data` argument, we will throw an `NullPointerException` here. I believe this should be fine as the previous code had the same "flaw." 

I just wanted to highlight this because one of your previous [comments](https://github.com/apache/kafka/pull/10747#discussion_r1058656097) left me a little confused.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java (hunk 9 above)
Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java:51]: Nevermind, think I've figured out the code changes needed to enable smooth rolling upgrades.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 11]
```diff
@@ -45,34 +47,85 @@ public void setIfUnset(final SerdeGetter getter) {
...
-            serializedKey = inner.serialize(topic, headers, data.oldValue);
+            throw new StreamsException("Both old and new values are null in ChangeSerializer, which is not allowed.");
         }
 
-        final ByteBuffer buf = ByteBuffer.allocate(serializedKey.length + NEWFLAG_SIZE);
-        buf.put(serializedKey);
-        buf.put((byte) (data.newValue != null ? 1 : 0));
-
         return buf.array();
     }
```
Comment (fqaiser94) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:135]: @mjsax Sorry for the delay, it took me a bit longer than expected to write up the KIP. 
Before I go submitting it formally, I wanted to share a [Google Doc version](https://docs.google.com/document/d/1RDulh8a01IeutPBQ92BM_8oWP-SadQZaRYnTqzkqtb8/edit?usp=sharing) with you in case you have any feedback. Feel free to drop any comments directly in the document. 
Otherwise, I will formally submit it in the next couple of weeks.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java (hunk 10 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: The implicit contract is, that `inner.serialize` can only return `null` if `data.oldValue` is `null`. Otherwise, the serializer has a bug. Hence, we should never hit a NPE here (or accept to hit it, to surface the bug in the serializer)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java (hunk 11 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:135]: Left view minor comments. Overall very well written. Looking forward to get the discussion started.

Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:135]: Thanks for your feedback!
Just FYI, I've created the KIP [here](https://cwiki.apache.org/confluence/x/P5VbDg). 
And started a discussion thread [here](https://lists.apache.org/thread/38rznfn748yx6gmz2b0ldohl45hq3b42).

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 12]
```diff
@@ -398,6 +398,12 @@ public class StreamsConfig extends AbstractConfig {
     @SuppressWarnings("WeakerAccess")
     public static final String UPGRADE_FROM_33 = "3.3";
 
+    /**
+     * Config value for parameter {@link #UPGRADE_FROM_CONFIG "upgrade.from"} for upgrading an application from version {@code 3.4.x}.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public static final String UPGRADE_FROM_34 = "3.4";
+
```
Comment (fqaiser94) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:406]: Not sure if it's appropriate to make this change in this PR. 
It _seems_ to me that the last step of the [Release Process](https://cwiki.apache.org/confluence/x/NqETAg) hasn't yet been completed for Kafka 3.4.0: 
> For feature releases, file a JIRA for updating compatibility/upgrade system tests to test the newly released version. Example PRs:
> - Broker and clients: https://github.com/apache/kafka/pull/12210
> - Streams: https://github.com/apache/kafka/pull/12209

**Review by fqaiser94 — COMMENTED** [review]
Updated PR based on minor feedback received during [KIP-904](https://cwiki.apache.org/confluence/x/P5VbDg) discussions.

File: docs/streams/upgrade-guide.html [hunk 13]
```diff
@@ -75,6 +75,20 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
...
+    <p>
+        Downgrading from 3.5.x or newer version to 3.4.x or older version needs special attention:
+        Since 3.5.0 release, Kafka Streams uses a new serialization format for repartition topics.
+        This means that older versions of Kafka Streams would not be able to recognize the bytes written by newer versions,
+        and hence it is harder to downgrade Kafka Streams with version 3.5.0 or newer to older versions in-flight. For
+        more details, please refer to <a href="https://cwiki.apache.org/confluence/x/P5VbDg">KIP-904</a>.
+
+        For a downgrade, first switch the config from <code>"upgrade.from"</code> to the version you are downgrading to.
+        This disables writing of the new serialization format in your application. It's important to wait in this state
+        for a few minutes to make sure that the application has finished processing any "in-flight" messages written
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: > for a few minutes 

We don't know how long it will take. I would just say "long enough" instead. -- In the end, the only way to tell would be to get the end-offsets for all repartition topics after the upgrade, and wait until the committed offsets become larger than those offsets.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 12 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:406]: That fine. Also 3.4 release was completed in the mean-time (but would have been ok before, too).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 14]
```diff
@@ -45,34 +49,87 @@ public void setIfUnset(final SerdeGetter getter) {
         }
     }
 
+    @SuppressWarnings("checkstyle:cyclomaticComplexity")
+    private boolean isUpgrade(final Map<String, ?> configs) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: Seems this method could be static?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 15]
```diff
@@ -45,34 +49,87 @@ public void setIfUnset(final SerdeGetter getter) {
...
+        final ByteBuffer buf;
+        if (!newValueIsNull && !oldValueIsNull) {
+            if (isUpgrade) {
                 throw new StreamsException("Both old and new values are not null (" + data.oldValue
-                    + " : " + data.newValue + ") in ChangeSerializer, which is not allowed.");
+                        + " : " + data.newValue + ") in ChangeSerializer, which is not allowed unless upgrading.");
+            } else {
+                final int capacity = UINT32_SIZE + newDataLength + oldDataLength + NEW_OLD_FLAG_SIZE;
+                buf = ByteBuffer.allocate(capacity);
+                ByteUtils.writeUnsignedInt(buf, newDataLength);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:118]: I don't think we gain anything using `writeUnsignedInt` that takes a `long`, because `newDataLength` is a regular int. -- and given that we set it as `byte[].length` we know that it cannot overflow anyway, because arrays in Java a limited to size `Integer.MAX_VALUE` -- it would be simpler/clearer to just use `writeVarint`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java [hunk 16]
```diff
@@ -46,16 +48,41 @@ public void setIfUnset(final SerdeGetter getter) {
...
+            final int oldDataLength = data.length - Integer.BYTES - newDataLength - NEW_OLD_FLAG_SIZE;
+            oldData = new byte[oldDataLength];
 
-        if (ByteBuffer.wrap(data).get(data.length - NEWFLAG_SIZE) != 0) {
-            return new Change<>(inner.deserialize(topic, headers, bytes), null);
+            buffer.get(newData);
+            buffer.get(oldData);
         } else {
-            return new Change<>(null, inner.deserialize(topic, headers, bytes));
+            throw new StreamsException("Encountered unknown byte value `" + newOldFlag + "` for oldNewFlag in in ChangedDeserializer.");
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedDeserializer.java]: nit: `in in`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 17]
```diff
@@ -45,34 +49,87 @@ public void setIfUnset(final SerdeGetter getter) {
...
      */
     @Override
     public byte[] serialize(final String topic, final Headers headers, final Change<T> data) {
-        final byte[] serializedKey;
-
-        // only one of the old / new values would be not null
-        if (data.newValue != null) {
-            if (data.oldValue != null) {
+        final boolean oldValueIsNull = data.oldValue == null;
+        final boolean newValueIsNull = data.newValue == null;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: Should we flip this around and say `newValueInNotNull`? -- This would avoid to have a negation`!newValueIsNull` later that is harder to read.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java [hunk 18]
```diff
@@ -45,34 +49,87 @@ public void setIfUnset(final SerdeGetter getter) {
...
+
+        final byte[] newData = inner.serialize(topic, headers, data.newValue);
+        final byte[] oldData = inner.serialize(topic, headers, data.oldValue);
+
+        final int newDataLength = newValueIsNull ? 0 : newData.length;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java]: If we switch to `newValueInNotNull`, we can just flip then/else to avoid an negation here.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java [hunk 19]
```diff
@@ -76,6 +79,48 @@ public boolean enableSendingOldValues(final boolean forceMaterialization) {
 
     private class KTableMapProcessor extends ContextualProcessor<K, Change<V>, K1, Change<V1>> {
 
+        private boolean isNotUpgrade;
+
+        @SuppressWarnings("checkstyle:cyclomaticComplexity")
+        private boolean isNotUpgrade(final Map<String, ?> configs) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java:85]: Can be `static`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/ChangedSerdeTest.java [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    @Test
+    public void shouldThrowErrorIfEncountersAnUnknownByteValueForOldNewFlag() {
+        final Change<String> data = new Change<>(null, nonNullOldValue);
+        final byte[] serialized = CHANGED_STRING_SERIALIZER.serialize(TOPIC, data);
+        assertThat(serialized, is(notNullValue()));
+
+        // mutate the serialized array to replace OLD_NEW_FLAG with an unsupported byte value
+        final ByteBuffer buffer = ByteBuffer.wrap(serialized);
+        buffer.position(serialized.length - 1);
+        buffer.put((byte) 3);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/ChangedSerdeTest.java]: Maybe use `-1` to be future prove? Otherwise we need to update this test when we use `3` as a flag.

**Review by mjsax — COMMENTED** [review]
Sorry for the delay in reviewing. Last few weeks were very busy. -- Still think we can make the 3.5 feature freeze deadline by next Wednesday.

Overall LGTM, just a couple of minor comments.

One thing: we should add a system test, to test the upgrade path. There is `StreamsUpgradeTest.java` and `streams_upgrade_test.py` (cf `test_rolling_upgrade_with_2_bounces`) as a starting point.

Might be best to do a similar thing as for FK-joins, and add a new test variation -- the tricky thing about the test would be, to ensure that the repartition topic is not empty when we do the bounce, so we should setup the test accordingly. -- If you want, we can also push this PR over the finish line, and add the system test in a follow-up PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java (hunk 19 above)
Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java:85]: Can't make this change: `Static declarations in inner classes are not supported at language level '8'`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java (hunk 15 above)
Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:118]: I agree that we gain little from using `writeUnsignedInt`. I mostly made this change to accommodate some [feedback](https://lists.apache.org/thread/2gov29vshy2x2mv3mhtcmhwq4w9627rc) from the KIP-904 discussion thread. 

I tried using the `writeVarint`/`readVarint` methods as per your suggestion but my unit test fails. 

If we still want something simpler, I would honestly just revert this [commit](https://github.com/apache/kafka/pull/10747/commits/46c48f26dfe85ec7970b5db1b410b7076d6d2bf8) and go back to using the `byteBuffer.putInt`/`byteBuffer.getInt` methods. I'm just not sure if i would need to go update the KIP document and inform folks before making this change; I don't think this is necessary for such a small detail but let me know your thoughts.

**Comment by fqaiser94** [comment]
> we should add a system test, to test the upgrade path.

I've started investigating this but it's unlikely that I will have this done before next Wednesday as I have an extremely busy next 2 weeks and this is looking a bit tricky (can't even get the existing system tests to pass for some reason currently).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java (hunk 15 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:118]: I would actually recommend to just remove it from the KIP -- it seem to be an implementation detail. \cc @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java (hunk 19 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableRepartitionMap.java:85]: Maybe the inner class would need to be `static`, too? Not sure. Also not too important.

File: docs/streams/upgrade-guide.html [hunk 21]
```diff
@@ -75,6 +75,20 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
...
+    <p>
+        Downgrading from 3.5.x or newer version to 3.4.x or older version needs special attention:
+        Since 3.5.0 release, Kafka Streams uses a new serialization format for repartition topics.
+        This means that older versions of Kafka Streams would not be able to recognize the bytes written by newer versions,
+        and hence it is harder to downgrade Kafka Streams with version 3.5.0 or newer to older versions in-flight. For
+        more details, please refer to <a href="https://cwiki.apache.org/confluence/x/P5VbDg">KIP-904</a>.
+
+        For a downgrade, first switch the config from <code>"upgrade.from"</code> to the version you are downgrading to.
+        This disables writing of the new serialization format in your application. It's important to wait in this state
+        log enough to make sure that the application has finished processing any "in-flight" messages written
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]:
```suggestion
        long enough to make sure that the application has finished processing any "in-flight" messages written
```

**Comment by mjsax** [comment]
Thanks for the quick turn around time! Merged. We can still do a follow up on the open question with regard to the serialization format. Using `Bytes.putInt()` should actually do, even if var-length encoding might be worth the trouble.

Hope we can get a system test earlier than later. If you are too busy, maybe somebody else can pick it up.

In any case: Thanks for the KIP and the PR!

**Comment by mjsax** [comment]
Follow up PR to user varint encoding: https://github.com/apache/kafka/pull/13533

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java (hunk 15 above)
Comment (fqaiser94) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/ChangedSerializer.java:118]: Just to close off this conversation, I've updated [KIP-904](https://cwiki.apache.org/confluence/x/P5VbDg) to replace the one reference to `UINT32` with just generic `INT`. If anyone reading this has any concerns, I suggest raising them in the follow-up PR [here](https://github.com/apache/kafka/pull/13533) rather than this in this PR which is already merged.

**Comment by fqaiser94** [comment]
Thanks for your support with this PR/KIP!

> Hope we can get a system test earlier than later. If you are too busy, maybe somebody else can pick it up.

Yea, sorry, I wont have the time to figure this out for at least the next couple of weeks :/ 
I've created a ticket for this work so we don't lose track of it here: https://issues.apache.org/jira/browse/KAFKA-14911
Feel free to add more details to the ticket. 
And if you can find someone else to work on this, please feel free to assign it to them.
